### PR TITLE
[#251] Refactor SVM to classify all outcomes

### PR DESCRIPTION
**Pull request by Samuel-Campbell**
[#251]

1) We have decided to use the SVM to classify ALL outcomes. Formerly, it only handled boolean outcomes but now we have decided to also include the integer ones. If the outcome for an integer is greater than 0 then we assume the asker received money (obtained his demand)

2) If the integer is greater than 0 then we send the facts to the regressor which will then make a prediction on $ amounts.

3) If the integer is 0 then we do not use the regressor. We simply return to the NLP that no money was awarded

4) The regressor is only used in winning cases because it was trained using biased data

5) Create new directory --> predictions 
    As the system gets more complex there will be many models to manage which should be done by this    module instead of the ML Controller

6) Currently the columns to replace booleans with integers are hardcoded simply because we do not have a wrapper for regression to handle it. GlobalPredictor.classifier_labels contains a dictionary where the key is the index of the column in the outcome matrix and the value is (outcome name, outcome type)
In the future: For every int_type --> run in regressor if boolean value is 1

### CODE REVIEW

**Comment by Samuel-Campbell** [comment]
INFO: Classifier results:

INFO: Column: additional_indemnity_date
INFO: Test accuracy: 88.19212808539027%
INFO: Precision: [ 0.91992551  0.84716901]
INFO: Recall: [ 0.84625268  0.92044403]
INFO: F1: [ 0.88155253  0.88228774]

INFO: Column: additional_indemnity_money
INFO: Test accuracy: 84.71203024238382%
INFO: Precision: [ 0.90034477  0.73544609]
INFO: Recall: [ 0.87715931  0.77862874]
INFO: F1: [ 0.88860083  0.75642161]

INFO: Column: declares_housing_inhabitable
INFO: Test accuracy: 99.98888147653992%
INFO: Precision: [ 0.99988881  0.        ]
INFO: Recall: [ 1.  0.]
INFO: F1: [ 0.9999444  0.       ]

INFO: Column: declares_resiliation_is_correct
INFO: Test accuracy: 92.37269290638203%
INFO: Precision: [ 0.96080454  0.6914378 ]
INFO: Recall: [ 0.95123819  0.73793103]
INFO: F1: [ 0.95599743  0.71392827]

INFO: Column: orders_expulsion
INFO: Test accuracy: 96.18634645319102%
INFO: Precision: [ 0.95135641  0.97725404]
INFO: Recall: [ 0.98393963  0.93204391]
INFO: F1: [ 0.96737373  0.95411371]

INFO: Column: orders_immediate_execution
INFO: Test accuracy: 88.0920613742495%
INFO: Precision: [ 0.93975703  0.76247906]
INFO: Recall: [ 0.88845186  0.86277483]
INFO: F1: [ 0.91338455  0.80953228]

INFO: Column: orders_resiliation
INFO: Test accuracy: 96.59773182121414%
INFO: Precision: [ 0.95607979  0.97962963]
INFO: Recall: [ 0.98478862  0.94175992]
INFO: F1: [ 0.97022188  0.96032158]

INFO: Column: orders_tenant_pay_first_of_month
INFO: Test accuracy: 99.2772959750945%
INFO: Precision: [ 0.99277296  0.        ]
INFO: Recall: [ 1.  0.]
INFO: F1: [ 0.99637337  0.        ]

INFO: Column: rejects_landlord_demand
INFO: Test accuracy: 97.36490993995997%
INFO: Precision: [ 0.97609607  0.84431138]
INFO: Recall: [ 0.99699144  0.40056818]
INFO: F1: [ 0.98643311  0.5433526 ]

INFO: Column: rejects_tenant_demand
INFO: Test accuracy: 97.16477651767845%
INFO: Precision: [ 0.9771027   0.69186047]
INFO: Recall: [ 0.99388908  0.37071651]
INFO: F1: [ 0.98542441  0.48275862]

INFO: Column: tenant_ordered_to_pay_landlord
INFO: Test accuracy: 70.2245941738937%
INFO: Precision: [ 0.82152975  0.59604876]
INFO: Recall: [ 0.64420585  0.78953229]
INFO: F1: [ 0.72214152  0.67928144]

INFO: Column: tenant_ordered_to_pay_landlord_legal_fees
INFO: Test accuracy: 75.91727818545696%
INFO: Precision: [ 0.88810055  0.60256095]
INFO: Recall: [ 0.73077565  0.81593865]
INFO: F1: [ 0.80179356  0.69320113]

**Comment by Samuel-Campbell** [comment]
Weights
---------------------
https://docs.google.com/spreadsheets/d/1l2t8phq7yZH0aGX56lmTfEMIB1MQfQcBAAvMDUdA_GQ/edit#gid=1691912729

File: src/ml_service/README.md [hunk 1]
```diff
@@ -166,9 +166,12 @@ export COMPOSE_FILE=ci
...
+    python main.py -train [data size | empty for all] --svm* --sf* --weights* --evaluate*
     python main.py -pre [number of files | empty for all]
-    python main.py -post
+    python main.py -post [number of files | empty for all]
+
 2- all binary files saved to data/binary
+
 3- all text files to data/cluster
+
 4- all raw_data to data/raw
```
Comment (dodoels) [inline src/ml_service/README.md:179]: all these command line documentation is broken in md format, please fix (not major issue)

File: src/ml_service/feature_extraction/post_processing/regex/regex_tagger.py [hunk 2]
```diff
@@ -35,15 +35,30 @@ def get_intent_index(self):
...
 
         outcomes_vector = []
         for i in range(len(self.regexes["regex_outcomes"])):
-            outcomes_vector.append((i, self.regexes["regex_outcomes"][i][0]))
+            name = self.regexes["regex_outcomes"][i][0]
+            if self.regexes["regex_outcomes"][i][2] == 'BOOLEAN':
+                data_type = 'bool'
+            else:
+                data_type = 'int'
+            outcomes_vector.append((i, name, data_type))
```
Comment (dodoels) [inline src/ml_service/feature_extraction/post_processing/regex/regex_tagger.py:61]: Not sure if using one for loop instead of 3 is a good idea, but I think these can be simplified

**Review by dodoels — COMMENTED** [review]
Looking good to me, and it is more scalable. Also update the file structure in README

**Comment by dodoels** [comment]
Unit test is failing plz fix

File: src/ml_service/model_training/classifier/multi_class_svm.py [hunk 3]
```diff
@@ -172,21 +174,17 @@ def reshape_dataset(self):
...
-            3.1) The data looks as such: [1, 1, 1, 0, 1, 0, 0, 1...]
-
-            3.2) We must create a new list with only the index of the columns where
+            2.2) We must create a new list with only the index of the columns where
                  there are values of '1'. This is necessary because the sklearn
                  algorithm expects this kind of input.
 
-            3.3) Example:        (transformation)
+            2.3) Example:        (transformation)
                 [1, 1, 0, 0, 1] ------------------> [0, 1, 4]
```
Comment (arekmano) [inline src/ml_service/model_training/classifier/multi_class_svm.py]: Its not clear to me why we are doing this transformation. Sklearn classifiers use 1-hot vectors AFAIK

File: src/ml_service/model_training/classifier/multi_class_svm.py [hunk 4]
```diff
@@ -200,19 +198,11 @@ def reshape_dataset(self):
         x_total = binarize(x_total, threshold=0)
 
         # --------------------2--------------------------
```
Comment (arekmano) [inline src/ml_service/model_training/classifier/multi_class_svm.py]: This method need to be split into multiple smaller ones

File: src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py [hunk 5]
```diff
@@ -88,8 +88,7 @@ def train(self):
         scaler = StandardScaler()
         self.model = TenantPaysLandlordRegressor.__create_pipeline(scaler, regressor)
         self.model.fit(X, Y)
-        self.dataset = None
-        self.save()
+        self.test()
```
Comment (arekmano) [inline src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py]: self.model.fit
and self.model.test both train the algorithm. Doing this will double training time

Comment (arekmano) [reply src/ml_service/model_training/regression/tenant_pays_landlord_regressor.py]: We should also make self.dataset = None, so that the dataset is not needlessly in memory

File: src/ml_service/model_training/classifier/multi_class_svm.py (hunk 3 above)
Comment (Samuel-Campbell) [reply src/ml_service/model_training/classifier/multi_class_svm.py]: the multioutput SVM takes the index of every outcome. If the index is part of the list then it palces a 1 in the column else a 0. It's a weird implementation i know. But we will eventually write our own wrapper

File: src/ml_service/feature_extraction/post_processing/regex/regex_tagger.py (hunk 2 above)
Comment (Samuel-Campbell) [reply src/ml_service/feature_extraction/post_processing/regex/regex_tagger.py:61]: how can they be simplified? every for loop has a different length

File: src/ml_service/README.md (hunk 1 above)
Comment (Samuel-Campbell) [reply src/ml_service/README.md:179]: fixed

**Review by dodoels — APPROVED** [review]
